Add a "Lonestar AI" to Fort Lonestar

abcdefg30 opened the pull request:
Makes use of the feature introduced in #10916.

pchote committed:
Add DummyAI trait for scripted maps.

obrakmann commented on mods/ra/maps/fort-lonestar/fort-lonestar-AI.lua:
```diff
@@ -0,0 +1,178 @@
+			AIBaseLocation[id] = barracks[1].Location + CVec.New(2, 1)
+			IdlingAIActors[id] = { }
+			IdlingAISupportActors[id] = { }
+			InitialInfatryProduction(id, InitialUnitsToBuild)
```
Typo: "Infatry"

obrakmann commented on mods/ra/maps/fort-lonestar/fort-lonestar.lua:
```diff
@@ -197,9 +197,13 @@ end
 		local player = Player.GetPlayer("Multi" ..i)
 		players[i] = player
+
+		if players[i] and players[i].Name == "Lonestar AI" then
```
Why not `and players[i].IsBot then`?

pchote replied:
I agree.  Checking against `IsBot` should be the minimum (because a real player may set their name to this), but if there are multiple AI types then checking `IsBot` and also the `Name` would be okay.

obrakmann commented:
Very nice, I like it. One thing I'd like to see is support of allied players, ie. when an ally is under attack, send a squad to help him. The AI appears a bit selfish otherwise ;)

In the test game with four AIs that I observed, two minor issues with Tanya crates occurred, which ultimately led to the game not finishing. The first was that the enemy picked up a Tanya crate outside of the range of the defending players, but didn't do anything with her. The defenders never went outside of their base to look for her, so the game never ended. The other issue was that the defending AI picked up a Tanya crate itself, but also didn't know what to do with her, she just idled. The first issue really only was an issue, because all the players were AI.

None of those thing should hold off this PR from getting merged, though. :+1:

abcdefg30 commented:
Updated.

I plan to enhance the AI anyway (somewhen...) later, so yeah at the moment it is "selfish". :D

Mailaender commented:
The crate hunting mechanic is impressive.

Bots don't seem get walls and don't build defenses. Is that on purpose?

pchote commented on mods/ra/maps/fort-lonestar/fort-lonestar-AI.lua:
```diff
@@ -0,0 +1,178 @@
+			return
+		end
+
+		local unit = Utils.Random((IdlingAIActors[id]))
```
Unnecessary `()` here?

pchote commented on mods/ra/maps/fort-lonestar/fort-lonestar-AI.lua:
```diff
@@ -0,0 +1,178 @@
+	end)
+end
+
+SellWalls = function(id)
```
I find this to be super confusing and unexpected.

abcdefg30 replied:
https://github.com/OpenRA/OpenRA/pull/10946#issuecomment-204829954

Phrohdoh commented:
I'd like to get this in soon so please consider [my fixup branch](https://github.com/Phrohdoh/OpenRA/commit/9f98c32c19c9cd9036a5b2e2ec74564e6d066b68) for this if @abcdefg30 doesn't return in the near future.

abcdefg30 committed:
Add an "Lonestar AI" to Fort Lonestar

abcdefg30 commented:
Rebased and updated.

I still think it's a bad idea to not sell the walls. We can of course try it without selling, but the AI behaves a lot better without them, IMHO. I just added a notification for now to indicate that the walls were sold and didn't disappear due to a bug.

Mailaender commented:
Needs a rebase.

abcdefg30 commented:
Rebased.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3bd5d502a3b2568874b0e3ab9e0704d601baa872...460248be9b303c83b45ea0a9e60053ffb7c6ae45)
